go/tools/astfmtgen/main.go

```diff
@@ -249,6 +247,50 @@ func (r *Rewriter) rewriteAstPrintf(cursor *astutil.Cursor, expr *ast.CallExpr)
 				Args: []ast.Expr{&ast.BasicLit{Value: `"%d"`, Kind: gotoken.STRING}, expr.Args[2+fieldnum]},
 			}
 			cursor.InsertBefore(r.rewriteLiteral(callexpr.X, "WriteString", call))
+		case 'n': // new directive for slices of AST nodes checked at code generation time
+			inputExpr := expr.Args[2+fieldnum]
+			inputType := r.pkg.TypesInfo.Types[inputExpr].Type
+			sliceType, ok := inputType.(*types.Slice)
+			if !ok {
+				panic("'%n' directive requires a slice")
+			}
+			if types.Implements(sliceType.Elem(), r.astExpr) {
+				// Fast path: input is []Expr
+				call := &ast.CallExpr{
+					Fun: &ast.SelectorExpr{
+						X:   callexpr.X,
+						Sel: &ast.Ident{Name: "formatExprs"},
+					},
+					Args: []ast.Expr{inputExpr},
+				}
+				cursor.InsertBefore(&ast.ExprStmt{X: call})
+				break
+			}
+			log.Printf("slow path for `n` directive with type %T", types.TypeString(inputType, noQualifier))
+			// Slow path: slice elements do not implement Expr
+			cursor.InsertBefore(&ast.ExprStmt{
+				X: &ast.CallExpr{
+					Fun: &ast.SelectorExpr{
+						X:   &ast.Ident{Name: "log"},
+						Sel: &ast.Ident{Name: "Printf"},
+					},
+					Args: []ast.Expr{
+						&ast.BasicLit{
+							Kind:  gotoken.STRING,
+							Value: strconv.Quote("slow path for %n with type %T"),
+						},
+						inputExpr,
+					},
+				},
+			})
```

Comment:
We really shouldn't be generating calls to `log.Print` in the fast formatter. What's the rationale here? Either the code is fast enough to run with the "slow path", in which case there's non need to warn about the slow path being taken, or it's too slow, in which case we should `panic` instead of generating.

Reply:
you and your logic